### PR TITLE
planner: add EXPLAIN FORMAT=JSON

### PR DESCRIPTION
**Pull request by nullnotnil**
### What problem does this PR solve?

Issue Number: Fixes https://github.com/pingcap/tidb/issues/20378

### What is changed and how it works?

What's Changed:

This implements `EXPLAIN FORMAT=JSON`.

Because the `operatorInfo` is built by concatenating strings at a much lower level (planner/core/explain.go:37 - dataAccessor interface), it is just returned as a string (not ideal). I have forked https://github.com/pingcap/tidb/issues/20520 to convert the delimiter to a `;`.

Here is an example:
```sql
*************************** 1. row ***************************
EXPLAIN: {
  "id": "HashJoin_17",
  "estRows": 12487.5,
  "task": "",
  "accessObject": "",
  "operatorInfo": "inner join, equal:[eq(test.t2.t1_id, test.t1.id)]",
  "children": [
    {
      "id": "TableReader_20(Build)",
      "estRows": 9990,
      "task": "",
      "accessObject": "",
      "operatorInfo": "data:Selection_19",
      "children": [
        {
          "id": "Selection_19",
          "estRows": 9990,
          "task": "",
          "accessObject": "",
          "operatorInfo": "not(isnull(test.t2.t1_id))",
          "children": [
            {
              "id": "TableFullScan_18",
              "estRows": 10000,
              "task": "",
              "accessObject": "table:t2",
              "operatorInfo": "keep order:false, stats:pseudo",
              "children": null
            }
          ]
        }
      ]
    },
    {
      "id": "TableReader_22(Probe)",
      "estRows": 10000,
      "task": "",
      "accessObject": "",
      "operatorInfo": "data:TableFullScan_21",
      "children": [
        {
          "id": "TableFullScan_21",
          "estRows": 10000,
          "task": "",
          "accessObject": "table:t1",
          "operatorInfo": "keep order:false, stats:pseudo",
          "children": null
        }
      ]
    }
  ]
}

1 row in set (0.00 sec)
```

### Related changes

- Parser PR: https://github.com/pingcap/parser/pull/1055

### Check List 

Tests 

- Unit tests

Side effects

- None

### Release note 

- The EXPLAIN statement now supports the optional FORMAT=JSON syntax.

### CODE REVIEW

**Comment by nullnotnil** [comment]
@breeswish @zz-jason may I ask your opinion on these two questions in the description?

**Comment by breeswish** [comment]
> (1) The golang JSON encoder will by default escape HTML characters. I think this creates ugly output in operatorInfo

Agree. Golang JSON encoder is escaped for HTML purpose and we don't need it.

> (2) The operatorInfo is built by concatenating strings at a much lower level (planner/core/explain.go:37 - dataAccessor interface)

We should change it (in future PRs). It is also the foundation of enabling more data to be visualized in TiDB Dashboard, for example, an interactive and nice plan view from slow query and statement page. @crazycs520 PTAL, we have talked about this previously.

**Comment by zz-jason** [comment]
> The best solution I have is to change concatenation between info items to use the semicolon character `;`, but this would be an incompatible change that also affects other explain formats.

@qw4990 do you have any idea?

**Comment by qw4990** [comment]
> The best solution I have is to change concatenation between info items to use the semicolon character ; , but this would be an incompatible change that also affects other explain formats.

I agree with it. Actually, I also meet this problem in [another project](https://github.com/qw4990/plan-change-capturer) where I want to parse and compare plans.

**Comment by nullnotnil** [comment]
This is now ready to review. The tests are just failing waiting on the parser PR to merge.

File: planner/core/common_plans.go [hunk 1]
```diff
@@ -1142,6 +1148,21 @@ func (e *Explain) getOperatorInfo(p Plan, id string) (string, string, string) {
 	return estRows, accessObject, operatorInfo
 }
 
+func (e *Explain) prepareJSONInfo(p PhysicalPlan) {
+	currentNode := &JSONOperatorRow{}
+	if err := e.explainPlanInJSONFormat(currentNode, p, "root", "", false); err != nil {
+		return
```
Comment (qw4990) [inline planner/core/common_plans.go:1154]: Should we return this error to users instead of ignoring it?

File: planner/core/common_plans.go [hunk 2]
```diff
@@ -1151,6 +1172,137 @@ func (e *Explain) prepareDotInfo(p PhysicalPlan) {
...
+	}
+	currentNode.ID = p.ExplainID().String() + driverSide
+	var (
+		tmpEstRows string
+		err        error
+	)
+	tmpEstRows, currentNode.AccessObject, currentNode.OperatorInfo = e.getOperatorInfo(p, currentNode.ID)
+	if currentNode.EstRows, err = strconv.ParseFloat(tmpEstRows, 64); err != nil {
+		currentNode.EstRows = 0
```
Comment (qw4990) [inline planner/core/common_plans.go:1197]: Should we set `currentNode.Task` to `taskType`?

Comment (qw4990) [reply planner/core/common_plans.go:1197]: I find the `task` field is empty in the result:
```
 {
  "id": "HashJoin_17",
  "estRows": 12487.5,
  "task": "",
  "accessObject": "",
  "operatorInfo": "inner join, equal:[eq(test.t2.t1_id, test.t1.id)]"
  ...
}
```

Comment (nullnotnil) [reply planner/core/common_plans.go:1197]: Yes :-) Thanks for spotting.

**Review by qw4990 — COMMENTED** [review]
~LGTM~

File: planner/core/plan_test.go [hunk 3]
```diff
@@ -448,3 +449,18 @@ func (s *testPlanNormalize) TestDecodePlanPerformance(c *C) {
...
+
+func (s *testPlanNormalize) TestJSONExplain(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("DROP TABLE IF EXISTS t1, t2")
+	tk.MustExec("CREATE TABLE t1 (id INT NOT NULL PRIMARY KEY, b INT)")
+	tk.MustExec("CREATE TABLE t2 (id INT NOT NULL PRIMARY KEY, t1_id INT)")
+
+	var js core.JSONOperatorRow
+	result := tk.MustQuery("EXPLAIN FORMAT=JSON SELECT t1.* FROM t1 INNER JOIN t2 ON t1.id=t2.t1_id")
```
Comment (qw4990) [inline planner/core/plan_test.go:461]: Please add more cases to cover more operators.

Comment (nullnotnil) [reply planner/core/plan_test.go:476]: @qw4990 I was worried about test stability as statistics and operator auto-naming can change. What I arrived at, was to use a regular expression. PTAL again, thx  :-)

**Comment by nullnotnil** [comment]
We can revisit this when https://github.com/pingcap/tidb/issues/20520 is complete.

**Pull request closed without merging**